### ISSUE TITLE
socket in close_wait status in Fluentd

### ISSUE THREAD

**Issue by havidarou** [wazuh/wazuh#3901]
|Wazuh version|Component|Install type|Platform|
|---|---|---|---|
| 3.9.0 | Fluent forwarder module | Agent | OS version |

It looks like the agent only detects that a connection no longer exist when it is sending messages.
Otherwise the socket will wait indefinitely.

To further analyze the impact of this issue we should answer:
- Does it block Fluentd? 
- Do descriptors leak?

Reference here: https://github.com/wazuh/wazuh/blob/master/src/wazuh_modules/wm_fluent.c#L670-L676

It may be related to/could be tackled at the same time as https://github.com/wazuh/wazuh/issues/3900.

**Comment by cristgl** [comment]
I haven't been able to reproduce this issue, the socket does not enter `CLOSE_WAIT` status although the other side of the TCP socket has been closed.

This has been the process followed to see the socket in `CLOSE_WAIT` status:
- Launch Fluentd and send data through the socket.
- Kill Fluentd and stop sending data. This should close the TCP socket and show its state as `CLOSE_WAIT`, as it is waiting for Fluentd to reconnect.
- When resending information through the socket, it should reconnect.

In order to see socket's status, two options:
1. Check `netstat` with `grep` by the port used:  `watch -d -n0 "netstat -a | grep 24224"`
2. Check `netstat` and get the PID with command `netstat -np | grep 24224`, check its state with `lsof`: `watch -d -n0 "lsof -np 8810 | grep 24224"`

**Comment by cristgl** [comment]
We haven't been able to reproduce this issue, however, we have checked that the functionality is correct. Whenever closing the socket which connects to `fluentd`, it gets reconnected and the new messages sent are being forwarded to `fluentd`.

**Comment by cristgl** [comment]
After a while, we were able to see `CLOSE_WAIT` status whenever the tag or the certificate did not match between server and client. 

In addition to that, when they are correct, they show the messages from below:
```
2019/09/11 15:37:39 fluent-forward: INFO: Connected to host 'demo' (192.168.0.132:24224)
2019/09/11 15:37:39 fluent-forward: INFO: Connected to 192.168.0.132:24224
2019/09/11 15:37:46 wazuh-modulesd:syscollector: INFO: Evaluation finished.
2019/09/11 15:37:46 fluent-forward: WARNING: Cannot send data to '192.168.0.132': Broken pipe (32). Reconnecting...
2019/09/11 15:37:46 fluent-forward: INFO: Connected to host 'demo' (192.168.0.132:24224)
2019/09/11 15:37:46 fluent-forward: INFO: Connected to 192.168.0.132:24224
2019/09/11 15:37:49 fluent-forward: WARNING: Cannot send data to '192.168.0.132': Broken pipe (32). Reconnecting...
2019/09/11 15:37:49 fluent-forward: INFO: Connected to host 'demo' (192.168.0.132:24224)
2019/09/11 15:37:49 fluent-forward: INFO: Connected to 192.168.0.132:24224
2019/09/11 15:38:07 fluent-forward: WARNING: Cannot send data to '192.168.0.132': Broken pipe (32). Reconnecting...
2019/09/11 15:38:07 fluent-forward: INFO: Connected to host 'demo' (192.168.0.132:24224)
2019/09/11 15:38:07 fluent-forward: INFO: Connected to 192.168.0.132:24224
2019/09/11 15:38:44 ossec-syscheckd: INFO: (6009): File integrity monitoring scan ended.
2019/09/11 15:40:20 fluent-forward: WARNING: Cannot send data to '192.168.0.132': Broken pipe (32). Reconnecting...
```

It is trying to reconnect in each message sent as `SSL_Write` is returning -1. Having a look at it.

**Comment by cristgl** [comment]
### Testing

Monitoring performance when sending a 1024 bytes message, checking the connection each time one of them is sent.

| O.S. | CPU | RAM |
| --- | --- | --- |
| Ubuntu 16.04 | 1 x 2.2GHz CPU | 1 GB |


### 1024 bytes string

| Total messages sent to UDP socket | Messages sent to fluentd | Time |
| :---: | :---: | :---: |
| 860790 | 2944 | 2 minutes |

- CPU performance

![cpu_n_1024](https://user-images.githubusercontent.com/14914658/65159155-f0398700-da33-11e9-9882-b127bc263005.jpg)

- RAM usage

![ram_n_1024](https://user-images.githubusercontent.com/14914658/65159165-f596d180-da33-11e9-8b7a-32dc9aedfbad.jpg)


### 32768 bytes string

| Total messages sent to UDP socket | Messages sent to fluentd | Time |
| :---: | :---: | :---: |
| 67603 | 3001 | 2 minutes |

- CPU performance

![cpu_n_32768](https://user-images.githubusercontent.com/14914658/65163379-1d3d6800-da3b-11e9-9c33-67ed4d6ddcac.jpg)

Monitoring performance when sending a 1024 bytes message, checking the connection before sending and when failing.


### 1024 bytes string

| Total messages sent to UDP socket | Messages sent to fluentd | Time |
| :---: | :---: | :---: |
| 786712 | 22462 | 2 minutes |

- CPU performance

![cpu_o_1024](https://user-images.githubusercontent.com/14914658/65164281-e5cfbb00-da3c-11e9-8c3a-ee4511728471.jpg)

- RAM usage

![ram_o_1024](https://user-images.githubusercontent.com/14914658/65164261-d8b2cc00-da3c-11e9-8aa8-c3ef9afae0b0.jpg)


### 32768 bytes string

| Total messages sent to UDP socket | Messages sent to fluentd | Time |
| :---: | :---: | :---: |
| 59355 | 13020 | 2 minutes |

- CPU performance
![cpu_o_32768](https://user-images.githubusercontent.com/14914658/65163321-0860d480-da3b-11e9-8cde-7db121e6c6c3.jpg)

- RAM usage
![ram_o_32768](https://user-images.githubusercontent.com/14914658/65163345-0d258880-da3b-11e9-9a4a-7b1fb7163e5c.jpg)

**Comment by cristgl** [comment]
With the new changes made to the code to send multiple messages to `fluentd` in branch [3901-send-chunks-fluentd](https://github.com/wazuh/wazuh/tree/3901-send-chunks-fluentd), the results when the maximum number of messages to send is 3, and we send 100 messages, is the next one:

```
2019-09-24 17:12:58.000000000 +0200 debug.test: ["0","1","2"]
2019-09-24 17:12:58.000000000 +0200 debug.test: ["3","4","5"]
2019-09-24 17:12:59.000000000 +0200 debug.test: ["6","7","8"]
2019-09-24 17:12:59.000000000 +0200 debug.test: ["9","10","11"]
2019-09-24 17:12:59.000000000 +0200 debug.test: ["12","13","14"]
2019-09-24 17:12:59.000000000 +0200 debug.test: ["15","16","17"]
2019-09-24 17:12:59.000000000 +0200 debug.test: ["18","19","20"]
2019-09-24 17:12:59.000000000 +0200 debug.test: ["21","22","23"]
2019-09-24 17:12:59.000000000 +0200 debug.test: ["24","25","26"]
2019-09-24 17:12:59.000000000 +0200 debug.test: ["27","28","29"]
2019-09-24 17:12:59.000000000 +0200 debug.test: ["30","31","32"]
2019-09-24 17:12:59.000000000 +0200 debug.test: ["33","34","35"]
2019-09-24 17:12:59.000000000 +0200 debug.test: ["36","37","38"]
2019-09-24 17:12:59.000000000 +0200 debug.test: ["39","40","41"]
2019-09-24 17:12:59.000000000 +0200 debug.test: ["42","43","44"]
2019-09-24 17:12:59.000000000 +0200 debug.test: ["45","46","47"]
2019-09-24 17:12:59.000000000 +0200 debug.test: ["48","49","50"]
2019-09-24 17:12:59.000000000 +0200 debug.test: ["51","52","53"]
2019-09-24 17:12:59.000000000 +0200 debug.test: ["54","55","56"]
2019-09-24 17:12:59.000000000 +0200 debug.test: ["57","58","59"]
2019-09-24 17:12:59.000000000 +0200 debug.test: ["60","61","62"]
2019-09-24 17:13:00.000000000 +0200 debug.test: ["63","64","65"]
2019-09-24 17:13:00.000000000 +0200 debug.test: ["66","67","68"]
2019-09-24 17:13:00.000000000 +0200 debug.test: ["69","70","71"]
2019-09-24 17:13:00.000000000 +0200 debug.test: ["72","73","74"]
2019-09-24 17:13:00.000000000 +0200 debug.test: ["75","76","77"]
2019-09-24 17:13:00.000000000 +0200 debug.test: ["78","79","80"]
2019-09-24 17:13:00.000000000 +0200 debug.test: ["81","82","83"]
2019-09-24 17:13:00.000000000 +0200 debug.test: ["84","85","86"]
2019-09-24 17:13:00.000000000 +0200 debug.test: ["87","88","89"]
2019-09-24 17:13:00.000000000 +0200 debug.test: ["90","91","92"]
2019-09-24 17:13:00.000000000 +0200 debug.test: ["93","94","95"]
2019-09-24 17:13:00.000000000 +0200 debug.test: ["96","97","98"]
2019-09-24 17:13:01.000000000 +0200 debug.test: ["99"]
```

**Comment by cristgl** [comment]
Testing
---------
Some testing has been made with the new changes to send chunks of logs to `fluentd`. At the configuration file, we have a localfile, monitoring the input to a file. This content is then forwarded to our `fluentd` UDP socket, /var/run/fluent.sock, which is read by the `fluentd` Wazuh module and forwarded to the server.

- Sending 10000 messages to `fluentd` with a sleep of 0.01 s between each message of a few bytes.
They are all sent, however, there are some inconsistencies as:
```
	["9521","9522","9523"]
	["9590","9591","9592"]
	["9524","9597","9598"]
```
But they are due to the forwarding from the localfile to the socket, as they are not shown in it.

**Comment by cristgl** [comment]
The following Valgrind reports where encountered:

```
==15303== Thread 2:
==15303== Conditional jump or move depends on uninitialised value(s)
==15303==    at 0x46EF02: wm_fluent_main (wm_fluent.c:118)
==15303==    by 0x58706B9: start_thread (pthread_create.c:333)
==15303==  Uninitialised value was created by a heap allocation
==15303==    at 0x4C2DE96: malloc (vg_replace_malloc.c:309)
==15303==    by 0x46EE16: wm_fluent_main (wm_fluent.c:104)
==15303==    by 0x58706B9: start_thread (pthread_create.c:333)
```

```
==27269== 131,072 bytes in 2 blocks are still reachable in loss record 225 of 225
==27269==    at 0x4C2FFAC: calloc (vg_replace_malloc.c:762)
==27269==    by 0x46EEDA: wm_fluent_main (wm_fluent.c:111)
==27269==    by 0x58706B9: start_thread (pthread_create.c:333)
```

Fixed in this commit: https://github.com/wazuh/wazuh/commit/c0e77f549fac9a200ee2d7421a6b527a8678664c

**Comment by vikman90** [comment]
Fixed in #4274.
